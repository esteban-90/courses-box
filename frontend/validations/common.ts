import { string } from 'yup'

/**
 * RegExp to match an username that is:
 * between five and eight alpha-numeric characters long
 * with zero or more numbers at the end,
 * optionally can start with a uppercase letter
 * and ends with only lowercase letters.
 */
export const usernameRegExp = /^(?=.{5,8})(?![a-z]\d?$)[A-Z]?[a-z]+\d*$/

/**
 * RegExp to match a password that has at least:
 * one lowercase letter,
 * one uppercase letter,
 * one digit,
 * one special character
 * and is between eight and ten characters long.
 */
const passwordRegExp = /(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[^A-Za-z0-9])(?=.{8,10})/

export const messages = {
  required: {
    field: 'This field is required',
  },

  invalid: {
    username: 'Username is not valid',
    email: 'Email is not valid',
    password: 'Password is not valid',
    passwordConfirmation: 'Passwords must match',
  },
}

export const password = string()
  .trim()
  .required(messages.required.field)
  .matches(passwordRegExp, messages.invalid.password)
