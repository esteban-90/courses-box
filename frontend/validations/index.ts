export * from './common'
export * from './login.schema'
export * from './register.schema'
