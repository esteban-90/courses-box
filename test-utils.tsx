import { FC, ReactElement, ReactNode } from 'react'
import { render, RenderOptions, RenderResult } from '@testing-library/react'
import { ThemeProvider } from '@emotion/react'
import { Themes } from '@/styles/themes'

const Wrapper: FC<{ children: ReactNode }> = ({ children }): JSX.Element => {
  return <ThemeProvider theme={Themes.light}>{children}</ThemeProvider>
}

const customRender = (ui: ReactElement, options?: RenderOptions): RenderResult => {
  return render(ui, { wrapper: Wrapper, ...options })
}

export * from '@testing-library/react'
export { customRender as render }

// 14
