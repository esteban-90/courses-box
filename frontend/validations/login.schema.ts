import { object, string } from 'yup'
import { messages, password, usernameRegExp } from '@/validations/common'

export const loginSchema = object().shape({
  identifier: string()
    .trim()
    .required(messages.required.field)
    .test('is-email', messages.invalid.email, (value) => {
      if (value) return value.includes('@') ? string().email().isValidSync(value) : true
      return true
    })
    .test('is-username', messages.invalid.username, (value) => {
      if (value) return !value.includes('@') ? string().matches(usernameRegExp).isValidSync(value) : true
      return true
    }),

  password,
})
