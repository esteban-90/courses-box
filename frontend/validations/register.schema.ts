import { object, ref, string } from 'yup'
import { messages, passwordRegExp, usernameRegExp } from '@/validations/common'

export const registerSchema = object().shape({
  username: string().trim().required(messages.required.field).matches(usernameRegExp, messages.invalid.username),
  email: string().trim().required(messages.required.field).email(messages.invalid.email),
  password: string().trim().required(messages.required.field).matches(passwordRegExp, messages.invalid.password),
  passwordConfirmation: string().oneOf([ref('password'), null], messages.invalid.passwordConfirmation),
})
