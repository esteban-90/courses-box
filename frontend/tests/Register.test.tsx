import userEvent from '@testing-library/user-event'
import { render, screen, act } from '@/test-utils'
import Register from '@/pages/register'
import { messages } from '@/validations'

describe('Register Page test cases', () => {
  it('Render check', () => {
    const { container } = render(<Register />)
    const registerForm = screen.getByTestId('form')

    expect(container).toMatchSnapshot()
    expect(registerForm).toHaveFormValues({ email: '', password: '', passwordConfirmation: '' })
  })

  it('Client validation check', async () => {
    render(<Register />)

    const submitButton = screen.getByRole('button', { name: 'Done' })
    await act(async () => await userEvent.click(submitButton))

    const requiredFeedbacks = screen.getAllByText(messages.required.field)
    expect(requiredFeedbacks).toHaveLength(3)

    const usernameField = screen.getByTestId('username')
    const emailField = screen.getByTestId('email')
    const passwordField = screen.getByTestId('password')
    const passConfirmField = screen.getByTestId('passwordConfirmation')

    await act(async () => {
      await userEvent.type(usernameField, 'test')
      await userEvent.type(emailField, 'test')
      await userEvent.type(passwordField, 'test')
    })

    const invalidUsernameFeedback = screen.getByText(messages.invalid.username)
    const invalidEmailFeedback = screen.getByText(messages.invalid.email)
    const invalidPasswordFeedback = screen.getByText(messages.invalid.password)

    expect(invalidUsernameFeedback).toBeInTheDocument()
    expect(invalidEmailFeedback).toBeInTheDocument()
    expect(invalidPasswordFeedback).toBeInTheDocument()

    await act(async () => {
      await userEvent.type(passwordField, 'test')
      await userEvent.type(passConfirmField, 'jest')
    })

    const invalidPassConfirmFeedback = screen.getByText(messages.invalid.passwordConfirmation)
    expect(invalidPassConfirmFeedback).toBeInTheDocument()

    await act(async () => {
      await userEvent.type(usernameField, 'Mytest12')
      await userEvent.type(emailField, 'test@example.com')
      await userEvent.type(passwordField, 'Test1234**')
      await userEvent.type(passConfirmField, 'Test1234**')
    })

    const validFeedbacks = screen.getAllByRole('alert')
    const [usernameValidFeedback, emailValidFeedback, passwordValidFeedback, passConfirmValidFeedback] = validFeedbacks

    expect(validFeedbacks).toHaveLength(4)
    expect(usernameValidFeedback).toMatchSnapshot()
    expect(emailValidFeedback).toMatchSnapshot()
    expect(passwordValidFeedback).toMatchSnapshot()
    expect(passConfirmValidFeedback).toMatchSnapshot()
  })
})
