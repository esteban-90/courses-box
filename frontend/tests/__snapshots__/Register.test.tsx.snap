// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Page test cases Client validation check 1`] = `
<span
  class="css-1reg724"
  role="alert"
/>
`;

exports[`Register Page test cases Client validation check 2`] = `
<span
  class="css-1reg724"
  role="alert"
>
  <span
    class="css-xjluvf"
  >
    Username is not valid
  </span>
</span>
`;

exports[`Register Page test cases Client validation check 3`] = `
<span
  class="css-1reg724"
  role="alert"
>
   
</span>
`;

exports[`Register Page test cases Client validation check 4`] = `
<span
  class="css-1reg724"
  role="alert"
>
  <span
    class="css-xjluvf"
  >
    Password is not valid
  </span>
</span>
`;

exports[`Register Page test cases Client validation check 5`] = `
<span
  class="css-1reg724"
  role="alert"
>
  <span
    class="css-xjluvf"
  >
    Passwords must match
  </span>
</span>
`;

exports[`Register Page test cases Render check 1`] = `
<div>
  <div
    class="css-19wldrt"
  >
    <a
      class="css-1nhvv70"
      href="/"
    >
      <header
        class="css-1l7s6s"
        size="3"
      >
        <span
          class="logo_short"
        >
          C8X
        </span>
        <span
          class="logo_full"
        >
          CoursesBox
        </span>
      </header>
    </a>
    <nav
      class="css-1sstk05"
    >
      <a
        class="css-1nz5qw1"
        href="/all"
      >
        All
      </a>
      <button
        class="css-1aqtdpz"
        size="2"
        title="Login"
      >
        <svg
          aria-label="Login"
          class="css-z3g3v6"
          fill="currentColor"
          height="1rem"
          href="/login"
          role="img"
          viewBox="0 0 1024 1024"
          width="1rem"
        >
          <defs>
            <style />
          </defs>
          <path
            d="M521.7 82c-152.5-.4-286.7 78.5-363.4 197.7-3.4 5.3.4 12.3 6.7 12.3h70.3c4.8 0 9.3-2.1 12.3-5.8 7-8.5 14.5-16.7 22.4-24.5 32.6-32.5 70.5-58.1 112.7-75.9 43.6-18.4 90-27.8 137.9-27.8 47.9 0 94.3 9.3 137.9 27.8 42.2 17.8 80.1 43.4 112.7 75.9 32.6 32.5 58.1 70.4 76 112.5C865.7 417.8 875 464.1 875 512c0 47.9-9.4 94.2-27.8 137.8-17.8 42.1-43.4 80-76 112.5s-70.5 58.1-112.7 75.9A352.8 352.8 0 01520.6 866c-47.9 0-94.3-9.4-137.9-27.8A353.84 353.84 0 01270 762.3c-7.9-7.9-15.3-16.1-22.4-24.5-3-3.7-7.6-5.8-12.3-5.8H165c-6.3 0-10.2 7-6.7 12.3C234.9 863.2 368.5 942 520.6 942c236.2 0 428-190.1 430.4-425.6C953.4 277.1 761.3 82.6 521.7 82zM395.02 624v-76h-314c-4.4 0-8-3.6-8-8v-56c0-4.4 3.6-8 8-8h314v-76c0-6.7 7.8-10.5 13-6.3l141.9 112a8 8 0 010 12.6l-141.9 112c-5.2 4.1-13 .4-13-6.3z"
          />
        </svg>
      </button>
      <button
        class="css-1aqtdpz"
        size="2"
        title="Moon"
      >
        <svg
          aria-label="Moon"
          class="css-z3g3v6"
          fill="currentColor"
          height="1rem"
          role="img"
          viewBox="0 0 16 16"
          width="1rem"
        >
          <path
            d="M6 .278a.768.768 0 01.08.858 7.208 7.208 0 00-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 01.81.316.733.733 0 01-.031.893A8.349 8.349 0 018.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 016 .278zM4.858 1.311A7.269 7.269 0 001.025 7.71c0 4.02 3.279 7.276 7.319 7.276a7.316 7.316 0 005.205-2.162c-.337.042-.68.063-1.029.063-4.61 0-8.343-3.714-8.343-8.29 0-1.167.242-2.278.681-3.286z"
          />
          <path
            d="M10.794 3.148a.217.217 0 01.412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 010 .412l-1.162.387a1.734 1.734 0 00-1.097 1.097l-.387 1.162a.217.217 0 01-.412 0l-.387-1.162A1.734 1.734 0 009.31 6.593l-1.162-.387a.217.217 0 010-.412l1.162-.387a1.734 1.734 0 001.097-1.097l.387-1.162zM13.863.099a.145.145 0 01.274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 010 .274l-.774.258a1.156 1.156 0 00-.732.732l-.258.774a.145.145 0 01-.274 0l-.258-.774a1.156 1.156 0 00-.732-.732l-.774-.258a.145.145 0 010-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"
          />
        </svg>
      </button>
    </nav>
    <label
      class="css-1m85ocu"
      height="7"
      width="20"
    >
      <span
        class="css-1grri41"
      />
      <div
        class="css-11gjszk"
      >
        <input
          class="css-1h6wv30"
          placeholder="Search"
        />
        <svg
          aria-label="Search"
          class="css-5n1pld"
          fill="currentColor"
          height="2rem"
          role="img"
          viewBox="0 0 16 16"
          width="2rem"
        >
          <path
            d="M11.742 10.344a6.5 6.5 0 10-1.397 1.398h-.001c.03.04.062.078.098.115l3.85 3.85a1 1 0 001.415-1.414l-3.85-3.85a1.007 1.007 0 00-.115-.1zM12 6.5a5.5 5.5 0 11-11 0 5.5 5.5 0 0111 0z"
          />
        </svg>
      </div>
      <span
        class="css-1reg724"
        role="alert"
      />
    </label>
    <main
      class="css-tee3o6"
    >
      <form
        data-testid="form"
        novalidate=""
      >
        <div
          class="css-gbwc6v"
        >
          <section
            class="css-1c469nw"
          >
            <h2>
              Register Page
            </h2>
            <h3>
               
            </h3>
            <label
              class="css-fjegpl"
              height="7"
              width="20"
            >
              <span
                class="css-1grri41"
              >
                Your username:
              </span>
              <div
                class="css-11gjszk"
              >
                <input
                  class="css-1h6wv30"
                  data-testid="username"
                  maxlength="8"
                  minlength="5"
                  name="username"
                  placeholder="..."
                  spellcheck="false"
                />
              </div>
              <span
                class="css-1reg724"
                role="alert"
              >
                 
              </span>
            </label>
            <label
              class="css-fjegpl"
              height="7"
              width="20"
            >
              <span
                class="css-1grri41"
              >
                Your email:
              </span>
              <div
                class="css-11gjszk"
              >
                <input
                  class="css-1h6wv30"
                  data-testid="email"
                  name="email"
                  placeholder="user@example.com"
                  type="email"
                />
              </div>
              <span
                class="css-1reg724"
                role="alert"
              >
                 
              </span>
            </label>
            <label
              class="css-fjegpl"
              height="7"
              width="20"
            >
              <span
                class="css-1grri41"
              >
                Your password:
              </span>
              <div
                class="css-11gjszk"
              >
                <input
                  class="css-1h6wv30"
                  data-testid="password"
                  maxlength="10"
                  minlength="8"
                  name="password"
                  placeholder="**********"
                  type="password"
                />
              </div>
              <span
                class="css-1reg724"
                role="alert"
              >
                 
              </span>
            </label>
            <label
              class="css-fjegpl"
              height="7"
              width="20"
            >
              <span
                class="css-1grri41"
              >
                Confirm your password:
              </span>
              <div
                class="css-11gjszk"
              >
                <input
                  class="css-1h6wv30"
                  data-testid="passwordConfirmation"
                  maxlength="10"
                  minlength="8"
                  name="passwordConfirmation"
                  placeholder="**********"
                  type="password"
                />
              </div>
              <span
                class="css-1reg724"
                role="alert"
              >
                 
              </span>
            </label>
            <button
              class="css-tp74ot"
              style="margin-top: 1rem;"
              type="submit"
            >
              Done
            </button>
            <a
              class="css-1ghpz66"
              href="/login"
              style="margin-top: 1rem; font-weight: bold;"
            >
              Access your account
            </a>
          </section>
        </div>
      </form>
    </main>
    <footer
      class="css-5ato5s"
    >
      <p>
        © 
        2022
         
        <a
          class="css-1nz5qw1"
          href="https://github.com/esteban-90"
          target="_blank"
        >
          Esteban V.M.
        </a>
         
      </p>
      <p>
        All rights reserved
      </p>
    </footer>
  </div>
</div>
`;
